src/EventListener/ControllerListener.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Setono\SyliusRedirectPlugin\EventListener;
+
+use Doctrine\Common\Persistence\ObjectManager;
+use Setono\SyliusRedirectPlugin\Model\RedirectInterface;
+use Setono\SyliusRedirectPlugin\Repository\RedirectRepositoryInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Event\FilterControllerEvent;
+
+final class ControllerListener
+{
+    /** @var RedirectRepositoryInterface */
+    private $redirectionRepository;
+    /** @var ObjectManager */
+    private $objectManager;
+    
+    /**
+     * ControllerListener constructor.
+     *
+     * @param RedirectRepositoryInterface $redirectionRepository
+     * @param ObjectManager               $objectManager
+     */
+    public function __construct(RedirectRepositoryInterface $redirectionRepository, ObjectManager $objectManager)
+    {
+        $this->redirectionRepository = $redirectionRepository;
+        $this->objectManager = $objectManager;
+    }
+    
+    /**
+     * @param FilterControllerEvent $event
+     *
+     * @throws \Doctrine\ORM\NonUniqueResultException
+     */
+    public function onKernelController(FilterControllerEvent $event): void
+    {
+        $request = $event->getRequest();
+        $pathInfo = $request->getPathInfo();
+        $baseUrl = $request->getBaseUrl();
+        
+        $redirect = $this->redirectionRepository->findEnabledBySource($pathInfo, false);
+        
+        if ($redirect instanceof RedirectInterface) {
+            $redirect->onAccess();
+            $this->objectManager->flush();
+            
+            $nextRedirect = $this->searchNextRedirection($redirect);
+            while ($nextRedirect instanceof RedirectInterface) {
+                $redirect = $nextRedirect;
+                $nextRedirect = $this->searchNextRedirection($redirect);
+            }
```

Comment:
You can rewrite these lines with a `do {} while();` instead:

```
do {
    $nextRedirect = $this->searchNextRedirection($redirect);
} while($nextRedirect && $redirect = $nextRedirect);
```

or something like that :)

Reply:
However I think this logic should be moved to the repository instead. You use this logic twice which is a very good indication of this :)